Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift

```diff
@@ -0,0 +1,467 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOHTTP1
+
+struct HTTPRequestStateMachine {
+    fileprivate enum State {
+        case initialized
+        case running(RequestState, ResponseState)
+        case finished
+
+        case failed(Error)
+    }
+
+    fileprivate enum RequestState {
+        enum ExpectedBody {
+            case length(Int)
+            case stream
+        }
+
+        enum ProducerControlState: Equatable {
+            case producing
+            case paused
+        }
+
+        case verifyRequest
+        case streaming(expectedBodyLength: Int?, sentBodyBytes: Int, producer: ProducerControlState)
+        case endSent
+    }
+
+    fileprivate enum ResponseState {
+        enum StreamControlState {
+            case downstreamHasDemand
+            case readEventPending
+            case waiting
+        }
+
+        case initialized
+        case receivingBody(StreamControlState)
+        case endReceived
+    }
+
+    enum Action {
```

Comment:
FWIW, in the past I've found it slightly awkward when debugging code which relies on a generic `Action` (rather than one action type per function on the state machine) since it's not clear at the call site what actions may occur as a result of prodding the state machine. I.e. `writabilityChanged` can only result in three of the 12 defined actions (`wait`, `resumeRequestBodyStream` and `pauseRequestBodyStream`) but that isn't clear when calling `writabilityChanged`. This is entirely personal preference though.